### ISSUE TITLE
Option to disable new stack prompt (if stack does not exist), just fail instead

### ISSUE THREAD

**Issue by ghostsquad** [pulumi/pulumi#9866]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

When running commands like `pulumi preview` against a stack that doesn't exist, I'm prompted to create the stack. I'd rather see this _fail_, but I don't have a way to handle this.

### Affected area/feature

CLI

**Comment by ghostsquad** [comment]
In fact, I know this would not be backwards compatible, but I feel this should be the _default_ behavior. It's actually a bit surprising to be asked to create a stack when I'm not running the `init` operation.

**Comment by siggimoo** [comment]
It does fail if you use the `--non-interactive` option. I just ran this on v3.34.1 in a directory with no stack:
```
% pulumi preview --non-interactive
error: no stack selected; please use `pulumi stack select` or `pulumi stack init` to choose one

% pulumi preview --stack nonexistent --non-interactive
error: no stack named 'nonexistent' found
```
In both cases the exit-code was non-zero.

**Comment by ghostsquad** [comment]
If I want to interact, but don't want to be prompted to make a new stack.

An empty stack configuration is not useful, so I'd rather it just fail.

**Comment by Frassle** [comment]
I think we could tackle this as part of the plans to add more typing to the configuration section. If a stack declared required configuration variables then we could either:
A) Fail to create new stacks in places like preview instead tell the user to run init and set the config values.
B) Allow the creating of a new stack but then interactively ask for all the required config values.

If a stack doesn't declare any required config values I see no harm in allowing the preview to create a new empty stack for ease of use in simple programs.

**Comment by ghostsquad** [comment]
A sounds nice

The problem is when something is misconfigured and you provide `--non-interactive --yes`. If I'm not explicitly creating a new stack, I want the operation to fail.

**Comment by siggimoo** [comment]
> The problem is when something is misconfigured and you provide `--non-interactive --yes`. If I'm not explicitly creating a new stack, I want the operation to fail.

Got it! Yeah, that could be a problem.

Considering this for a moment, I think it represents an issue of overall CLI philosophy. Creating a stack is a separate command from previewing or applying or anything else. Prompting users to create one is a courtesy but not a necessity.

For normal, interactive CLI operation I think it's fine to continue as is. There aren't any surprises as the user is given notice and a chance to act (or abort). But for non-interactive operation, I think such courtesy behaviors should be avoided. Each command should do its one job and nothing else. If things aren't adequately prepared, that's the user's problem. Let them sort it out. If they're using non-interactive mode, there shouldn't be any hand-holding.

**Comment by ghostsquad** [comment]
It's not even _just_ about user interactivity. I want to _craft_ the user experience that the Pulumi folks will not know (nor should they). They are providing the tools, I need to use the tools. It's the same reason to wrap `pulumi` in `Task`. To that end, even in an `interactive` situation, I don't want a user to be prompted to create a stack, as that convenience shortcut is not an approved workflow. It's not well supported by Pulumi either, as documented here (see the `template` section):

https://www.pulumi.com/docs/reference/pulumi-yaml/

**Comment by siggimoo** [comment]
I really have no opinion on the interactive mode. I use it only in very simple, experimental situations. But I agree that tools like Pulumi should not force usage patterns. Suggest and recommend, sure, but don't compel.

**Comment by iwahbe** [comment]
Could you elaborate on your use case? This would be a breaking change to the UX.

**Comment by ghostsquad** [comment]
First, in automation, my only options are `--non-interactive --yes` for options.

What's less surprising in the following situation:

Facts:
- Stack `poduction` doesn't exist (notice the misspelling)
- There's a misconfiguration where I'm trying to bring up a different stack (that does exist), but the automation filled in the blanks incorrectly.

Result 1

`pulumi up` results in `no stack by the name "poduction" exists`.

Result 2


`pulumi up` results in `would you like to create stack "poduction"?`.

---

Both results may cause me to ask "why doesn't the stack exist?".

But neither scenario is me explicitly asking for a stack to be created.

So if my expectation is that the stack exists, `up` should either succeed or fail. It should not _also_ create the stack.

---

I would be happy with just a flag like `--no-create-on-missing-stack` for backwards compatibility.

**Comment by iwahbe** [comment]
@ghostsquad I still don't understand your use case. I can think of two situations:
1. `pulumi` is being run from a script: If a valid stack is not selected, or if the `--stack` flag is passed and points to a non-existent stack then the command errors. You are not prompted. I include this case since you mention automation, but we don't prompt when `--non-interactive` is passed. 
2. You are running `pulumi` manually: I understand that pressing`^C` is annoying, but it sounds like it doesn't block any workflow. If you want to confirm a stack exists before running `pulumi up` you can use `pulumi stack select`, which will error on a missing stack. Do the semantics of `pulumi stack select <stack> && pulumi up` do what you want us to do?

**Comment by ghostsquad** [comment]
It's annoying. Nailed it. I'm glad you understand.

In automation, if I pass `--yes` and `--non-interactive` does it

1. create a stack, and skip further prompts (agreeing when necessary)?

2. Or not create a stack but agree to other prompts?

If 1, then, I still need the behavior described... (select && up)

If 2, does this not surprise you? Wouldn't you expect `--yes` to apply to all prompts? If so, probably would want to do that... Then we would be left with result 1.

**Comment by iwahbe** [comment]
If you pass `--non-interactive --yes` then `pulumi` will error if the stack doesn't exist. `--yes` approves yes/no questions, but the create a stack prompts you for the stack name.

**Comment by ghostsquad** [comment]
It doesn't prompt with the stack name if I've provided `--stack`. It is a yes/no at that point.

**Comment by ghostsquad** [comment]
Oh I apologize .. no I don't use `select`. I use `--stack`. Which allows me to be very very very sure what's happening, and avoids race conditions.

**Comment by siggimoo** [comment]
> It doesn't prompt with the stack name if I've provided `--stack`. It is a yes/no at that point.

I agree. If the user is being asked any sort of proceed/abort question then I would expect the following behavior:
- `<no options>` prompts for user input
- `--yes` automatically proceeds
- `--non-interactive` automatically aborts
- `--non-interactive --yes` automatically proceeds

**Comment by ghostsquad** [comment]
> > It doesn't prompt with the stack name if I've provided `--stack`. It is a yes/no at that point.
> 
> I agree. If the user is being asked any sort of proceed/abort question then I would expect the following behavior:
> - `<no options>` prompts for user input
> - `--yes` automatically proceeds
> - `--non-interactive` automatically aborts
> - `--non-interactive --yes` automatically proceeds
> 

Right so... There's no way to skip the yes/no of creating a new stack and fail instead with a "no stack error"

**Comment by siggimoo** [comment]
> Right so... There's no way to skip the yes/no of creating a new stack and fail instead with a "no stack error"

I'm not sure I follow what you're saying here. The problem that I see right now is the inconsistent implementation of `--yes`. `pulumi preview` currently fails on a nonexistent stack:

```
% pulumi preview --stack nonexistent --non-interactive
error: no stack named 'nonexistent' found
```

`pulumi up` also fails under the same circumstances but with slightly different command syntax:

```
% pulumi up --stack nonexistent --non-interactive --yes
error: no stack named 'nonexistent' found
```

However, you cannot put `--yes` on `preview` and you're required to put it on `up`:

```
% pulumi preview --stack nonexistent --non-interactive --yes
...
An error occurred: unknown flag: --yes

% pulumi up --stack nonexistent --non-interactive      
error: --yes must be passed in to proceed when running in non-interactive mode
```

This is what I think needs to be fixed.

**Comment by ghostsquad** [comment]
Oh wow. Ok so, actually the problem is not only in the interaction but the actual behavior too.

So, I'll be honest here, I didn't actually verify that `--stack nonexistent --non-interactive --yes` didn't create a stack. The signifiers I got from running with `--stack nonexistent` separately from `--non-interactive --yes` lead me to write this issue, thinking that `--yes` would answer yes to the `would you like to create this stack?` prompt.

So at this point, I see several problems. Inconsistent behavior with the `--yes` and the undesired "convenience" function, when in interactive mode.

**Comment by siggimoo** [comment]
> So, I'll be honest here, I didn't actually verify that `--stack nonexistent --non-interactive --yes` didn't create a stack. The signifiers I got from running with `--stack nonexistent` separately from `--non-interactive --yes` lead me to write this issue, thinking that `--yes` would answer yes to the `would you like to create this stack?` prompt.

This right here demonstrates why this issue should be addressed. Your intuition is completely reasonable. The behavior of the command is not.

**Comment by ghostsquad** [comment]
The "convenience" of asking for a stack on any other command still doesn't feel right, so that hard decision to not break backwards compatibility requires that this behavior not change. In that case, adding an option, and maybe a configuration value to allow this "feature" in both interactive and non-interactive modes to be disabled seems appropriate.

What would be absolutely fantastic would be if this option could be set in `Pulumi.yml`, that would be one less thing that a wrapping script needs to handle then.

**Comment by zhangyoufu** [comment]
I'd like an option or environment variable to automatically create stack in non-interactive mode. This is useful for GitLab dynamic environment management, so I don't have to write `pulumi stack init XXX || true; pulumi up`

**Comment by zhangyoufu** [comment]
> Would `pulumi stack select --stack XXX --create` work for you?

Thank you! It works.

I used to think pulumi save the current selected stack in the state backend. Apparently I was wrong. It's stored in `~/.pulumi/workspaces` or overridden by environment variable `PULUMI_STACK`.
